File: Leetcode/leetcode-0144-BinaryTreePreorderTraversal-[tree]-[easy]/Solution.iterative1.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function (root) {
    let result = [];
    let stack = [];
    if (root === null) {
        return result;
    }

    let cur = root;
    while (stack.length !== 0 || cur !== null) {
        while (cur !== null) {
            result.push(cur.val);
            stack.push(cur);
            cur = cur.left;
        }

        cur = stack.pop();
        cur = cur.right;
    }

    return result;
};

File: Leetcode/leetcode-0144-BinaryTreePreorderTraversal-[tree]-[easy]/Solution.iterative1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        TreeNode cur = root;
        while (!stack.isEmpty() || cur != null) {
            while (cur != null) {
                result.add(cur.val);
                stack.push(cur);
                cur = cur.left;
            }
            
            cur = stack.pop();
            cur = cur.right;
        }
        
        return result;
    }
}